File: src/generic/correspondence_estimation.cpp
#include "searchwrapper.h"
#include <pcl/registration/correspondence_estimation.h>

#define CorrespondenceEstimation_ptr pcl::registration::CorrespondenceEstimation<_PointT, _PointT>::Ptr
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define KdTree_ptr pcl::KdTreeFLANN<_PointT>::Ptr

PCLIMP(CorrespondenceEstimation_ptr *, CorrespondenceEstimation, new)()
{
  return new CorrespondenceEstimation_ptr(new pcl::registration::CorrespondenceEstimation<_PointT, _PointT>());
}

PCLIMP(void, CorrespondenceEstimation, delete)(CorrespondenceEstimation_ptr *self)
{
  delete self;
}

PCLIMP(void, CorrespondenceEstimation, setInputSource)(CorrespondenceEstimation_ptr *self, PointCloud_ptr *cloud)
{
  (*self)->setInputSource(*cloud);
}

PCLIMP(void, CorrespondenceEstimation, setInputTarget)(CorrespondenceEstimation_ptr *self, PointCloud_ptr *cloud)
{
  (*self)->setInputTarget(*cloud);
}

PCLIMP(void, CorrespondenceEstimation, setIndicesSource)(CorrespondenceEstimation_ptr *self, Indices_ptr *indices)
{
  (*self)->setIndicesSource(*indices);
}
  
PCLIMP(void, CorrespondenceEstimation, setIndicesTarget)(CorrespondenceEstimation_ptr *self, Indices_ptr *indices)
{
  (*self)->setIndicesTarget(*indices);
}

PCLIMP(void, CorrespondenceEstimation, setSearchMethodSource)(CorrespondenceEstimation_ptr *self, KdTree_ptr *kdtree, bool force_no_recompute = false)
{
  boost::shared_ptr<KdTreeSearchWrapper<_PointT> > wrapper(new KdTreeSearchWrapper<_PointT>(*kdtree));
  (*self)->setSearchMethodSource(wrapper, force_no_recompute);
}

PCLIMP(void, CorrespondenceEstimation, setSearchMethodTarget)(CorrespondenceEstimation_ptr *self, KdTree_ptr *kdtree, bool force_no_recompute = false)
{
  boost::shared_ptr<KdTreeSearchWrapper<_PointT> > wrapper(new KdTreeSearchWrapper<_PointT>(*kdtree));
  (*self)->setSearchMethodTarget(wrapper, force_no_recompute);
}

PCLIMP(void, CorrespondenceEstimation, determineCorrespondences)(
  CorrespondenceEstimation_ptr *self,
  Correspondences_ptr *correspondences,
  double max_distance =std::numeric_limits<double>::max()
)
{
  (*self)->determineCorrespondences(**correspondences, max_distance);
}

PCLIMP(void, CorrespondenceEstimation, determineReciprocalCorrespondences)(
  CorrespondenceEstimation_ptr *self,
  Correspondences_ptr *correspondences,
  double max_distance = std::numeric_limits<double>::max()
)
{
  (*self)->determineReciprocalCorrespondences(**correspondences, max_distance);
}

#undef CorrespondenceEstimation_ptr
#undef PointCloud_ptr
#undef KdTree_ptr

File: CorrespondenceEstimation.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local CorrespondenceEstimation = torch.class('pcl.CorrespondenceEstimation', pcl)

local func_by_type = {}

local function init()

  local CorrespondenceEstimation_method_names = {
    'new',
    'delete',
    'setInputSource',
    'setInputTarget',
    'setIndicesSource',
    'setIndicesTarget',
    'setSearchMethodSource',
    'setSearchMethodTarget',
    'determineCorrespondences',
    'determineReciprocalCorrespondences'
  }
  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_CorrespondenceEstimation_TYPE_KEY_", CorrespondenceEstimation_method_names, v)
  end
  func_by_type[pcl.FPFHSignature33] = utils.create_typed_methods("pcl_CorrespondenceEstimation_TYPE_KEY_", CorrespondenceEstimation_method_names, 'FPFHSignature33')
end

init()

function CorrespondenceEstimation:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = func_by_type[self.pointType]
  self.o = self.f.new()
end

function CorrespondenceEstimation:cdata()
  return self.o
end

function CorrespondenceEstimation:setInputSource(cloud)
  self.f.setInputSource(self.o, cloud:cdata())
end

function CorrespondenceEstimation:setInputTarget(cloud)
  self.f.setInputTarget(self.o, cloud:data())
end

function CorrespondenceEstimation:setIndicesSource(indices, no_recompute)
  self.f.setIndicesSource(self.o, indices:cdata(), no_recompute or false)
end

function CorrespondenceEstimation:setIndicesTarget(indices, no_recompute)
  self.f.setIndicesTarget(self.o, indices:cdata(), no_recompute or false)
end

function CorrespondenceEstimation:setSearchMethodSource(kdtree)
  self.f.setSearchMethodSource(self.o, kdtree:cdata())
end

function CorrespondenceEstimation:setSearchMethodTarget(kdtree)
  self.f.setSearchMethodTarget(self.o, kdtree:cdata())
end

function CorrespondenceEstimation:determineCorrespondences(correspondences, max_distance)
  correspondences = correspondences or pcl.Correspondences()
  self.f.determineCorrespondences(self.o, correspondences:cdata(), max_distance or pcl.range.double.max)
  return correspondences
end

function CorrespondenceEstimation:determineReciprocalCorrespondences(correspondences, max_distance)
  correspondences = correspondences or pcl.Correspondences()
  self.f.determineReciprocalCorrespondences(self.o, correspondences:cdata(), max_distance or pcl.range.double.max)
  return correspondences
end
